### PR TITLE
Set timezone for group column timestamp type

### LINKED ISSUE #13534: Simple query fails with `column types must match schema types`
### Describe the bug

Found in our production system using datafusion internals, but reproducible in `datafusion-cli`:

```sql
SELECT
    'foo' AS text,
    arrow_cast('2024-01-01T00:00:00Z'::timestamptz, 'Timestamp(Microsecond, Some("UTC"))') AS ts
GROUP BY ts, text
```

Gives:

```
Arrow error: Invalid argument error: column types must match schema types, expected Timestamp(Microsecond, Some("UTC")) but found Timestamp(Microsecond, None) at column index 0
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

I bisected it with this script:

```shell
pushd datafusion-cli && cargo run --bin datafusion-cli -- --file ../q.sql && popd
```

And then:

```shell
git checkout c0ca4b4  # current HEAD
git bisect start
git bisect bad
git checkout 88f58bf # 43.0, I manually verified it is okay
git bisect good
git bisect run ./test.sh  # the above script
```

The result points to ecc04d4af85a29111a1598e615350fea84e60fcb / https://github.com/apache/datafusion/pull/13457

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #13534.

### CODE REVIEW

File: datafusion/physical-plan/src/aggregates/group_values/multi_group_by/primitive.rs [hunk 1]
```diff
@@ -35,6 +36,7 @@ use std::sync::Arc;
 /// `NULLABLE`: if the data can contain any nulls
 #[derive(Debug)]
 pub struct PrimitiveGroupValueBuilder<T: ArrowPrimitiveType, const NULLABLE: bool> {
+    data_type: DataType,
```
Comment (alamb) [inline datafusion/physical-plan/src/aggregates/group_values/multi_group_by/primitive.rs:39]: I double checked and the single group by column implementation does the same thing
https://github.com/apache/datafusion/blob/c0ca4b4e449e07c3bcd6f3593fa31dd31ed5e0c5/datafusion/physical-plan/src/aggregates/group_values/single_group_by/primitive.rs#L84

👍

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211  -- this looks good to me

**Review by adriangb — APPROVED** [review]
Looks great to me thank you for the quick fix!

**Comment by alamb** [comment]
Thanks again @jayzhan211 and @adriangb

**Pull request merged**